### PR TITLE
refactor(metadata): pre-refactors for migrating feature states to static metadata [part 3/12]

### PR DESCRIPTION
**Pull request by glevco**
Depends on https://github.com/HathorNetwork/hathor-core/pull/1014

---

### Motivation

Continuing with the static metadata refactor introduced in PRs https://github.com/HathorNetwork/hathor-core/pull/1013 and https://github.com/HathorNetwork/hathor-core/pull/1014, the next step is to move the `feature_states` attribute from metadata to static metadata.

This was initially implemented in PR https://github.com/HathorNetwork/hathor-core/pull/1015, which turned out to be too large, so it was divided into 3 PRs: this one, https://github.com/HathorNetwork/hathor-core/pull/1017, and https://github.com/HathorNetwork/hathor-core/pull/1018. The latter is the one that actually moves the `feature_states` attribute, and the first two PRs perform simple refactors that were either necessary or small improvements.

Therefore, this PR simply performs some pre-refactors to debloat the main PR.

### Acceptance Criteria

- **IMPORTANT:** move `verify_aux_pow()` verification from `verify_without_storage()` to `verify()`.
- Rename `FeatureService.get_bits_description()` to `get_feature_infos()`, and `FeatureDescription` to `FeatureInfo`.
- Change `FeatureService._get_ancestor_at_height()` to assert that a block exists in the height index when it must.
- Update the `FeatureState` enum to be a unique string enum, like `Feature` (this will be important for serialization), and implement `is_active()`.
- Change static metadata dataclasses to `BaseModel`s instead, which will be necessary for (de)serialization.
- Other smaller improvements.

### Checklist

- [x] If you are requesting a merge into `master`, confirm this code is production-ready and can be included in future releases as soon as it gets merged

### CODE REVIEW

File: tests/feature_activation/test_feature_service.py [hunk 1]
```diff
@@ -445,10 +394,8 @@ def test_get_state_from_active(storage: TransactionStorage, block_height: int) -
 
 
 @pytest.mark.parametrize('block_height', [16, 17, 18, 19])
-def test_caching_mechanism(storage: TransactionStorage, block_height: int) -> None:
```
Comment (glevco) [inline tests/feature_activation/test_feature_service.py:448]: This test was removed because in part 5 the caching mechanism is removed, as it's not necessary anymore.

File: hathor/transaction/static_metadata.py [hunk 2]
```diff
@@ -68,7 +62,6 @@ def from_bytes(cls, data: bytes, *, target: 'BaseTransaction') -> 'VertexStaticM
         raise NotImplementedError
 
 
-@dataclass(slots=True, frozen=True, kw_only=True)
```
Comment (msbrogli) [inline hathor/transaction/static_metadata.py:71]: Why are you removing this line?

File: hathor/transaction/static_metadata.py [hunk 3]
```diff
@@ -181,7 +174,6 @@ def _get_previous_feature_activation_bit_counts(
         return parent_block.static_metadata.feature_activation_bit_counts
 
 
-@dataclass(slots=True, frozen=True, kw_only=True)
```
Comment (msbrogli) [inline hathor/transaction/static_metadata.py:184]: Why are you removing this line?

File: hathor/transaction/static_metadata.py [hunk 4]
```diff
@@ -26,16 +24,16 @@
...
+from hathor.util import json_loadb
+from hathor.utils.pydantic import BaseModel
 
 if TYPE_CHECKING:
     from hathor.conf.settings import HathorSettings
     from hathor.transaction import BaseTransaction, Block, Transaction
     from hathor.transaction.storage import TransactionStorage
 
 
-@dataclass(slots=True, frozen=True, kw_only=True)
```
Comment (msbrogli) [inline hathor/transaction/static_metadata.py:37]: Why are you removing this line?

File: hathor/verification/verification_service.py [hunk 5]
```diff
@@ -274,7 +275,6 @@ def _verify_without_storage_block(self, block: Block) -> None:
         self._verify_without_storage_base_block(block)
 
     def _verify_without_storage_merge_mined_block(self, block: MergeMinedBlock) -> None:
-        self.verifiers.merge_mined_block.verify_aux_pow(block)
```
Comment (msbrogli) [inline hathor/verification/verification_service.py:277]: Why can't we run this verification without storage? This change seems completely unrelated to the other refactors in this PR.

File: hathor/builder/builder.py
Comment (msbrogli) [inline hathor/builder/builder.py:1]: I still don't get why you're doing this refactor. It might become clear when reviewing the next PRs.

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/transaction/static_metadata.py (hunk 2 above)
Comment (glevco) [reply hathor/transaction/static_metadata.py:71]: I mentioned it in the Acceptance Criteria:

> Change static metadata dataclasses to `BaseModel`s instead, which will be necessary for (de)serialization.

Specifically, I'm now serializing the `Feature` and `FeatureState` enums. Before, I was using `dataclasses.asdict()` which serializes enums as objects that cannot be converted to json, instead of using their raw values directly. Then, I changed those classes to `BaseModel` which are equivalent and provide the `json_dumpb()` method, so I don't have to do this conversion manually.

File: hathor/transaction/static_metadata.py (hunk 3 above)
Comment (glevco) [reply hathor/transaction/static_metadata.py:184]: Please see https://github.com/HathorNetwork/hathor-core/pull/1016#discussion_r1729346188

File: hathor/transaction/static_metadata.py (hunk 4 above)
Comment (glevco) [reply hathor/transaction/static_metadata.py:37]: Please see https://github.com/HathorNetwork/hathor-core/pull/1016#discussion_r1729346188

File: hathor/verification/verification_service.py (hunk 5 above)
Comment (glevco) [reply hathor/verification/verification_service.py:277]: As explained in slack, this was wrong as this verification does in fact use the storage.

File: hathor/builder/builder.py
Comment (glevco) [reply hathor/builder/builder.py:1]: Yes, in the next PR the `FeatureService` becomes just a utility class (that can even be removed later, and converted to functions), and not a singleton anymore. Therefore, it's simpler to use it by passing `settings` as argument instead of `settings.FEATURE_ACTIVATION`

File: hathor/verification/verification_service.py (hunk 5 above)
Comment (msbrogli) [reply hathor/verification/verification_service.py:277]: Got it. Thanks!

File: hathor/transaction/static_metadata.py (hunk 2 above)
Comment (msbrogli) [reply hathor/transaction/static_metadata.py:71]: Is `BaseModel` mutable? If yes, I don't think it's worth losing immutability because of a serialization method.

File: tests/tx/test_verification.py [hunk 6]
```diff
@@ -342,8 +342,6 @@ def test_merge_mined_block_verify_without_storage(self) -> None:
         verify_data_wrapped = Mock(wraps=self.verifiers.block.verify_data)
         verify_sigops_output_wrapped = Mock(wraps=self.verifiers.vertex.verify_sigops_output)
 
-        verify_aux_pow_wrapped = Mock(wraps=self.verifiers.merge_mined_block.verify_aux_pow)
```
Comment (msbrogli) [inline tests/tx/test_verification.py:345]: Don't you have to add this mock and `assert_called_once()` to the moved where the call was moved to?

File: hathor/transaction/static_metadata.py (hunk 2 above)
Comment (glevco) [reply hathor/transaction/static_metadata.py:71]: It's immutable, as configured here: https://github.com/HathorNetwork/hathor-core/blob/da32f8092a1ae038046864a4edc6e448b9e1b811/hathor/utils/pydantic.py#L33

File: tests/tx/test_verification.py (hunk 6 above)
Comment (glevco) [reply tests/tx/test_verification.py:345]: It's already there, since `verify` and `validate_full` called `verify_without_storage`, this was already being tested: https://github.com/HathorNetwork/hathor-core/blob/da32f8092a1ae038046864a4edc6e448b9e1b811/tests/tx/test_verification.py#L409

File: hathor/transaction/static_metadata.py (hunk 2 above)
Comment (msbrogli) [reply hathor/transaction/static_metadata.py:71]: Got it. Thanks for clarifying.

File: tests/tx/test_verification.py (hunk 6 above)
Comment (msbrogli) [reply tests/tx/test_verification.py:345]: It is actually the `verify_aux_pow` (not the `verify_pow`). But I checked and it's there. Thanks!

**Review by msbrogli — APPROVED** [review]

File: hathor/builder/builder.py
Comment (msbrogli) [reply hathor/builder/builder.py:1]: Got it. I'd rather see this change there but it seems harmless to be merged in this PR.

**Review by jansegre — APPROVED** [review]
